feat: add `array/base/reshape`

headlessNode opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   adds `array/base/reshape` including tests, benchmarks and documentation.

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/20

## Questions

> Any questions for reviewers of this pull request?

I'm not sure if the function signature is as expected, it was not mentioned in the issue.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

headlessNode committed:
feat: add array/base/reshape

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: passed
---

headlessNode commented on lib/node_modules/@stdlib/array/base/reshape/docs/types/index.d.ts:
```diff
@@ -0,0 +1,57 @@
+* var out = reshape( x, [ 2, 3 ], [ 3, 2 ], false );
+* // returns [ [ 1, 2 ], [ 3, 4 ], [ 5, 6 ] ]
+*/
+declare function reshape<T = unknown>( x: ArrayND<T>, fromShape: Shape, toShape: Shape, colexicographic: boolean ): Array<T>;
```
```suggestion
declare function reshape<T = unknown>( x: ArrayND<T>, fromShape: Shape, toShape: Shape, colexicographic: boolean ): ArrayND<T>;
```

headlessNode committed:
fix: apply suggestions from code review

Signed-off-by: Muhammad Haris <101793258+headlessNode@users.noreply.github.com>

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/README.md:
```diff
@@ -0,0 +1,117 @@
+
+## Notes
+
+-   The function assumes that the `fromShape` and the `toShape` have the same number of dimensions.
```
Why is this restriction necessary? Rather than the same number of dimensions, what we need to assume is that the two shapes have the same number of elements.

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/lib/main.js:
```diff
@@ -0,0 +1,99 @@
+	}
+
+	out = [];
+	for ( j = 0; j < S; j++ ) {
```
Why do you need a separate variable `j` when you've already got `i` which is used above?

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/lib/main.js:
```diff
@@ -0,0 +1,99 @@
+
+	S = shape[ dim ];
+	if ( dim === ndims - 1 ) {
+		return slice( array, index, index + S );
```
The use of `slice` was clever here in order to create the subarrays.

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/README.md:
```diff
@@ -0,0 +1,117 @@
+
+# reshape
+
+> Reshape an n-dimensional nested array into another m-dimensional having a desired shape.
```
```suggestion
> Reshape a nested array into another nested array having a desired shape.
```

Applies here and elsewhere.

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/README.md:
````diff
@@ -0,0 +1,117 @@
+// returns [ [ 1, 2 ], [ 3, 4 ], [ 5, 6 ] ]
+```
+
+To reshape in colexicographic order, provide a third argument equal to `true`.
````
```suggestion
To reshape in colexicographic order, set the `colexicographic` argument to `true`.
```

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/benchmark/benchmark.js:
```diff
@@ -0,0 +1,206 @@
+
+// MAIN //
+
+bench( pkg+':ndims=2,size=100,lexicographic', function benchmark( b ) {
```
Use factorization to ensure that these benchmarks all involve nested arrays having the same number of elements. This allows consumers to more readily infer expected perf as the number of dimensions increases.

```
Factorization of 100:

1, 2, 4, 5, 10, 20, 25, 50, 100

dims 2: 10, 10
dims 3: 4, 5, 5
dims 4: 5, 2, 5, 2
dims 5: 5, 2, 1, 5, 2
```

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/docs/types/index.d.ts:
```diff
@@ -0,0 +1,57 @@
+*
+* ## Notes
+*
+* -   The function assumes that `fromShape` and `toShape` have same the number of dimensions.
```
Same comment. elements not dimensions.

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/docs/types/index.d.ts:
```diff
@@ -0,0 +1,57 @@
+* @returns output m-dimensional nested array
+*
+* @example
+* var reshape = require( '@stdlib/array/base/reshape' );
+*
```
```suggestion
```

We shouldn't be requiring the package in TypeScript definitions.

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/docs/types/test.ts:
```diff
@@ -0,0 +1,97 @@
+{
+	const x = [ [ 1, 2 ], [ 3, 4 ] ];
+
+	reshape<number>( x, [ 2, 2 ], [ 4, 1 ], false ); // $ExpectType number[]
```
Is it necessary to explicitly use `<number>` here? 

Same thing for `<string>` below. I would think that the compiler is smart enough to infer.

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/docs/types/test.ts:
```diff
@@ -0,0 +1,97 @@
+{
+	const x = [ [ 1, 2 ], [ 3, 4 ] ];
+
+	reshape( x, [ 2, 2 ], [ 4, 1], '' ); // $ExpectError
```
[ 4, 1 ] spacing.

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/lib/main.js:
```diff
@@ -0,0 +1,99 @@
+* @returns {Array} output m-dimensional nested array
+*
+* @example
+* var reshape = require( '@stdlib/array/base/reshape' );
+*
```
```suggestion
```

Why are you requiring the package here?

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/test/test.js:
```diff
@@ -0,0 +1,337 @@
+	t.end();
+});
+
+tape( 'the function returns a reshaped nested array (2d, lexicographic)', function test( t ) {
```
All your tests here and below assume the same number of dimensions. You need to include additional tests for reshaping arrays to nested arrays having different numbers of dimensions. E.g., 2d to 1d, 2d to 3d, etc. You don't need to be completely exhaustive, as the number of dimensions is unbounded, but you should at least demonstrate this for each array rank (2d, 3d, 4d, 5d, etc), which you've tested below.

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/test/test.js:
```diff
@@ -0,0 +1,337 @@
+	x = [
+		[
+			[
+				[ 1, 2, 3, 4]
```
```suggestion
				[ 1, 2, 3, 4 ]
```

Use consistent spacing.

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/test/test.js:
```diff
@@ -0,0 +1,337 @@
+	x = [
+		[
+			[
+				[ 1, 2, 3, 4]
```
```suggestion
				[ 1, 2, 3, 4 ]
```

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: passed
  - task: lint_repl_help
    status: passed
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: passed
  - task: lint_typescript_tests
    status: passed
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
refactor: apply suggestions from code review

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: passed
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode committed:
test: add additional tests

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: na
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

headlessNode commented:
@kgryte Changes applied. Please check.

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/README.md:
````diff
@@ -0,0 +1,117 @@
+
+var out = reshape( x, [ 2, 3 ], [ 3, 2 ], false );
+// returns [ [ 1, 2 ], [ 3, 4 ], [ 5, 6 ] ]
+```
````
@headlessNode Would you mind adding a list of parameter descriptions here?

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/README.md:
```diff
@@ -0,0 +1,117 @@
+
+## Notes
+
+-   The function assumes that the `fromShape` and the `toShape` have the same number of elements.
```
```suggestion
-   The function assumes that `fromShape` and `toShape` have the same number of elements.
```

Making consistent with note in src JSDoc comment.

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/lib/main.js:
```diff
@@ -0,0 +1,96 @@
+* // returns [ [ 1, 2 ], [ 3, 4 ], [ 5, 6 ] ]
+*/
+function reshape( x, fromShape, toShape, colexicographic ) {
+	var f;
+
+	f = flatten( x, fromShape, colexicographic );
```
```suggestion
	var f = flatten( x, fromShape, colexicographic );
```

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/README.md:
````diff
@@ -0,0 +1,117 @@
+var out = reshape( x, [ 2, 3 ], [ 3, 2 ], false );
+// returns [ [ 1, 2 ], [ 3, 4 ], [ 5, 6 ] ]
+```
+
````
```suggestion
-   **x**: input array.
-   **fromShape**: input array shape.
-   **toShape**: output array shape.
-   **colexicographic**: boolean indicating whether to reshape the array in colexicographic order.

```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/README.md:
```diff
@@ -0,0 +1,121 @@
+
+## Notes
+
+-   The function assumes that `fromShape` and `toShape` have the same number of elements.
```
```suggestion
-   The function assumes that `fromShape` and `toShape` specify arrays having the same number of elements.
```

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/docs/types/index.d.ts:
```diff
@@ -0,0 +1,55 @@
+*
+* ## Notes
+*
+* -   The function assumes that `fromShape` and `toShape` have same the number of elements.
```
```suggestion
* -   The function assumes that `fromShape` and `toShape` specify arrays have same the number of elements.
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/README.md:
```diff
@@ -0,0 +1,121 @@
+
+## Notes
+
+-   The function assumes that `fromShape` and `toShape` specify arrays having the same number of elements.
```
```suggestion
-   The function assumes that `fromShape` and `toShape` describe arrays having the same number of elements.
```

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/docs/types/index.d.ts:
```diff
@@ -0,0 +1,55 @@
+*
+* ## Notes
+*
+* -   The function assumes that `fromShape` and `toShape` specify arrays have same the number of elements.
```
```suggestion
* -   The function assumes that `fromShape` and `toShape` describe arrays have same the number of elements.
```

kgryte committed:
Apply suggestions from code review

Signed-off-by: Athan <kgryte@gmail.com>

kgryte reviewed (dismissed):
LGTM

kgryte commented on lib/node_modules/@stdlib/array/base/reshape/README.md:
````diff
@@ -0,0 +1,121 @@
+var out = reshape( x, [ 2, 3 ], [ 3, 2 ], false );
+// returns [ [ 1, 2 ], [ 3, 4 ], [ 5, 6 ] ]
+```
+-   **x**: input array.
````
```suggestion

-   **x**: input array.
```

kgryte committed:
style: fix missing blank line

Signed-off-by: Athan <kgryte@gmail.com>
